sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,147 @@
+            request.requestId
+          );
+        } catch (e) {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received an error from request`,
+            request.requestId
+          );
+          responseError = e as RestError;
+          if (responseError.name !== "RestError") {
+            throw responseError;
+          }
+          response = responseError.response;
+        }
+
+        retryPolicyLogger.info(
+          `Retry ${retryCount}: Processing ${strategies.length} retry strategies.`
+        );
+
+        strategiesLoop: for (const strategy of strategies) {
+          const strategyLogger = strategy.logger || retryPolicyLogger;
+          strategyLogger.info(`Retry ${retryCount}: Processing retry strategy ${strategy.name}.`);
+
+          const information: RetryInformation = {
+            retryCount,
+            response,
+            responseError
+          };
+
+          if (information.retryCount >= options.maxRetries ?? DEFAULT_MAX_RETRIES) {
+            strategyLogger.info(
+              `Maximum retries reached. Returning the last received response, or throwing the last received error.`
+            );
+            if (response !== undefined) {
+              return response;
+            }
+            throw responseError;
+          }
+
+          let modifiers: RetryModifiers;
+          try {
+            modifiers = strategy.retry(information);
+          } catch (e) {
+            const error = e as Error;
+            if (error.name === "SkipRetryError") {
+              strategyLogger.error(`Retry ${retryCount}: Skipped. ${error.message}`);
+              continue strategiesLoop;
+            }
+            throw e;
+          }
+
+          if (request.abortSignal?.aborted) {
+            strategyLogger.error(`Retry ${retryCount}: Request aborted.`);
+            const abortError = new AbortError();
+            throw abortError;
+          }
+
+          const { throwError, retryAfterInMs, redirectTo } = modifiers;
+
+          if (throwError) {
+            strategyLogger.error(
+              `Retry ${retryCount}: Retry strategy ${strategy.name} throws error:`,
+              throwError
+            );
+            throw throwError;
+          }
+
+          if (retryAfterInMs) {
+            strategyLogger.info(
+              `Retry ${retryCount}: Retry strategy ${strategy.name} retries after ${retryAfterInMs}`
+            );
+            await delay(retryAfterInMs);
+            continue retryRequest;
+          }
+
+          if (redirectTo) {
+            strategyLogger.info(
+              `Retry ${retryCount}: Retry strategy ${strategy.name} redirects to ${redirectTo}`
+            );
+            request.url = redirectTo;
+            continue retryRequest;
+          }
+        }
+        retryPolicyLogger.info(`No retry strategy left. Returning the last received response.`);
+        if (response) {
```

Comment:
I'm a little worried about this since the contract of this method is a successful return *must* return a PipelineResponse but we have this suspicious if check which implies it could return `undefined`

Reply:
This if is within the retries loop. If there’s a response, we return it, otherwise we continue the loop (going again through the strategies). In the worst case scenario we reach the max retries.

Reply:
I finally understand this, since we're still in the while(true), so max retries logic up above will handle errors